### ISSUE TITLE
@QuarkusIntegrationTest + Docker: Allow Customization

### ISSUE THREAD

**Issue by schulzp** [quarkusio/quarkus#17560]
## Description

Thanks for introducing `@QuarkusIntegrationTest`! Currently, we use a two-staged process to obtain our native image:

1. Build graalvm-based image with `JAVA_TOOL_OPTIONS=-agentlib:native-image-agent=config-merge-dir=/var/native-image`, run integration tests against that image mounting `src/main/resources/META-INF/native-image` as volume under `/var/native-image`.
2. Build native image, which picks up `META-INF/native-image` by default, and re-run integration tests

The first run captures all uses of resources and reflection for any of our test scenarios, the second run immediately considers them to build a native image. This is extremely handy when working with libraries that have not been optimized for native usage yet. However, `QuarkusIntegrationTestExtension` does not allow such customization.

Since `@NativeImageTest` has been announced to be deprecated at some point, I'd highly appreciate it if `@QuarkusIntegrationTest` was as flexible as `@NativeImageTest`.

## Implementation ideas

Would it be feasible to work with **SPIs** to customize `DockerContainerLauncher` or even replace it with a custom implementation based on [Test Containers](https://www.testcontainers.org/)?

**Comment by geoand** [comment]
We can certainly make `DockerContainerLauncher` customizable.
I assume you would like to be able to add argument to the `docker run...` command?

Furthermore, how did `@NativeImageTest` satisfy the use case mentioned above?

Thanks

**Comment by schulzp** [comment]
> I assume you would like to be able to add argument to the `docker run...` command?

Yes. Yet, an SPI which would allow using Test Containers to interact with docker would simplify things, for example, volume mounts.

> Furthermore, how did `@NativeImageTest` satisfy the use case mentioned above?

The associated `NativeTestExtension` allows specifying `native.image.path` which I use to specify a path to a shell script instead of a binary. That shell script does, what the new `DockerContainerLauncher` does: it maps arguments passed to it (application properties) to environment variables and passes them on when calling `docker run...`. The docker image to be run is extracted from a custom non-application property, injected by a custom `DockerTestResource`. Additionally, I implement Quarkus' SPI `io.quarkus.test.common.NativeImageStartedNotifier` to test `localhost:<dynamic-port>/health`. All in all, quite a few workarounds.

Thanks!

**Comment by geoand** [comment]
Very interesting information, thanks.

So if we introduced an SPI that would essentially pass in the container image to use, the ports, the log file and expect a process to be launched with that info, would that be enough?

**Comment by schulzp** [comment]
Yes, that sounds workable! As for the initialization, I'd expect the FQN of the container image in combination with a parameter object of a new type `ArtifactLauncherContext` that provides a reference to `org.eclipse.microprofile.config.Config` as well as the log file path and anything else that might be relevant in future.

I'm not so sure about "expect[ing] a process to be launched". The SPI should require `start() throws IOException` and `Closable|AutoCloseable.stop() throws IOException` as required by `ArtifactLauncher`.  Besides that, the calling side should not expect anything.

However, we'd need a back-channel through which the launcher can inform the test extension about the actual `test.url`, since with docker + dynamic port mapping, we do not know that upfront. So instead of/in addition to `io.quarkus.test.common.ArtifactLauncher#listensOnSsl` maybe a new method `List<uri>|Map<protocol, uri> getTestUris()` would provide more flexibility.
